test_conformance/extensions/cl_khr_command_buffer/command_buffer_queue_substitution.cpp

```diff
@@ -0,0 +1,349 @@
+          ADD_PROP(CL_QUEUE_PROFILING_ENABLE)
+        };
+
+        for ( auto && prop : props )
+        {
+          check_property(prop);
+          test_error(error, "clCreateCommandQueue failed");
+          if (queue_with_prop!=nullptr)
+              return CL_SUCCESS;
+        }
+
+        return CL_INVALID_QUEUE_PROPERTIES;
+    }
+
+    //--------------------------------------------------------------------------
+    cl_int SetUp(int elements) override
+    {
+        // By default command queue is created without properties,
+        // if test requires queue with properties default queue must be replaced.
+        if (properties_use_requested)
+        {
+            // due to the skip condition
+            queue = nullptr;
+            if ( CreateCommandQueueWithProperties(queue) != CL_SUCCESS )
+              return CL_SUCCESS;
+        }
+
+        return BasicCommandBufferTest::SetUp(elements);
+    }
+
+    //--------------------------------------------------------------------------
+    cl_int Run() override
+    {
+        cl_int error = CL_SUCCESS;
+        cl_command_queue_properties cqp;
+        error = clGetCommandQueueInfo(queue, CL_QUEUE_PROPERTIES, sizeof(cqp), &cqp, NULL);
+        test_error(error, "clGetCommandQueueInfo failed");
+
+        if(properties_use_requested && (cqp & CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE))
+        {
+          out_of_order_queue = true;
+        }
+
+        // record command buffer with primary queue
+        error = RecordCommandBuffer();
+        test_error(error, "RecordCommandBuffer failed");
+
+        // create substitute queue
+        cl_command_queue new_queue = nullptr;
+        if (properties_use_requested)
+        {
+            error = CreateCommandQueueWithProperties(new_queue);
+            test_error(error, "CreateCommandQueueWithProperties failed");
+        }
+        else
+        {
+            const cl_command_queue_properties queue_properties = 0;
+            new_queue = clCreateCommandQueue
+                (context, device, queue_properties, &error);
+            test_error(error, "clCreateCommandQueue failed");
+        }
+
+        if (simultaneous_use)
+        {
+          // enque simultaneous command-buffers with substitute queue
+          error = RunSimultaneous(new_queue);
+          test_error(error, "RunSimultaneous failed");
+        }
+        else
+        {
+          // enque single command-buffer with substitute queue
+          error = RunSingle(new_queue);
+          test_error(error, "RunSingle failed");
+        }
+
+        clReleaseCommandQueue(new_queue);
+        if (properties_use_requested)
+        {
+            clReleaseCommandQueue(queue);
+        }
+        return CL_SUCCESS;
+    }
+
+    //--------------------------------------------------------------------------
+    cl_int RecordCommandBuffer()
+    {
+      cl_int error = CL_SUCCESS;
+
+      if(out_of_order_queue)
```

Comment:
Could this condition be `if (!simultaneous_use)` instead?

Thanks to the sync point, the below command-buffer would have the correct ordering if a queue was either in-order or out-of-order. And could then remove
```
  if (!out_of_order_queue)
      {
        error = clEnqueueFillBuffer(q, in_mem, &pattern_pri, sizeof(cl_int),
                                    0, data_size(), 0, nullptr, nullptr);
        test_error(error, "clEnqueueFillBuffer failed");
      }
```
From `RunSingle()`

Reply:
Done, there was one more malfunction of queue with properties test scenario. As result of this correction flag "out_of_order_queue" became unnecessary.